geode-core/src/main/java/org/apache/geode/internal/net/ByteBufferSharingImpl.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.apache.geode.internal.net;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * An {@link AutoCloseable} meant to be acquired in a try-with-resources statement. The resource (a
+ * {@link ByteBuffer}) is available (for reading and modification) in the scope of the
+ * try-with-resources. Collaborates with {@link ByteBufferReferencing} to ensure final dereference
+ * returns the {@link ByteBuffer} to the pool.
+ */
+class ByteBufferSharingImpl implements ByteBufferSharing {
+
+  private final ByteBuffer buffer;
+  private final Lock lock;
+  private final ByteBufferReferencing referencing;
+
+  /**
+   * This constructor is for use only by the owner of the shared resource (a {@link ByteBuffer}).
+   * Furthermore, this constructor is only for use in constructing the very first {@link
+   * ByteBufferSharingImpl} for the resource. Subsequent instances must be constructed via the copy
+   * constructor.
+   *
+   * A resource owner calls this constructor and retains the reference and never hands it out.
+   * Instead, the owner hands out copies, constructed via the copy constructor.
+   *
+   * This constructor acquires no lock and does not modify the reference count through the {@link
+   * ByteBufferReferencing} (passed in here.)
+   */
+  ByteBufferSharingImpl(final ByteBuffer buffer,
+      final ByteBufferReferencing referencing) {
+    this(buffer, new ReentrantLock(), referencing);
+  }
+
+  /**
+   * This constructor is for use only by the owner of the shared resource. It's used for handing out
+   * references to the shared resource. So it does reference counting and also acquires a lock.
+   *
+   * Resource owners will typically return the result of this constructor, to a caller. That caller
+   * binds that reference to a variable in a try-with-resources statement and relies on the
+   * AutoCloseable protocol to invoke close() on the object at the end of the block.
+   */
+  ByteBufferSharingImpl(final ByteBufferSharingImpl other) {
```

Comment:
Always creating a new instance at the start of try-with-resources is not needed. In the case of this object I think it would be okay for the same instance to be used by multiple threads concurrently and each one to call close concurrently on the same instance. If so this would address Jakes concern about object allocation.

Reply:
This is addressed in the latest commit. This copy-constructor was replaced by this method:

```
ByteBuffer alias()
```

The resource-owner returns the result of invoking `alias()` everywhere it hands out an alias. No more construction!